### PR TITLE
[Feature]: Save both best and final models on available apps (NVFLARE)

### PR DESCRIPTION
**Pull request by virginiafdez**
# Description

This PR  modifies the nvflare main controller and constants to allow to save a 'best' model in addition to final one. The best model is saved by looking into the output validation metrics. 

The user specifies the following parameters:
```
  "BEST_MODEL_METRIC":  example: "VAL-F1-SCORE",
  "BEST_MODEL_METRIC_MINIMIZE": false,
```

Then the job recipe updates the configuration. With this, the controller stores the values of this metric provided by the client and uses the last one to decide whether the model is getting better or know (per-client averaging).
Based on that, it saves a best model state dict.

## Linked Issues

#673 

Fixes #

## Checklist

- [x] Follows the project's coding conventions and style guide
- [ ] Updates documentation
- [] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

## Type of Change

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make -C docs/ docs`.

## Testing

`make -C fl-tutorials run-tutorial TUTORIAL=3d_spleen_segmentation FL_BACKEND=nvflare 2>&1`

- [x] Quick tests passed locally by running `make unit-test`.
- [ ] Integration tests pass (if applicable) by running `make integration-test`.

## Additional Notes

Missing tests:
- X-Ray (already implemented, but not tested)
- Diffusion model (already implemented, but not tested)

🚨  The issue #673 shouldn't be closed because Flower needs a separate update.

### CODE REVIEW

**Comment by atriaybagur** [comment]
I've done major updates to scatter_and_gather.py in https://github.com/londonaicentre/FLIP/pull/710 for BDMS fixes - I'll set this PR to draft while we merge, then I'll work on resolving any conflicts with this branch.

**Review by atriaybagur — CHANGES_REQUESTED** [review]
Thanks for tackling this — best-model saving is genuinely useful to have, and the config-key UX (`BEST_MODEL_METRIC` / `BEST_MODEL_METRIC_MINIMIZE`) is the right shape.

My main recommendation is to build it on NVFLARE's **built-in** best-model machinery rather than hand-rolled tracking inside the FLIP `ScatterAndGather`: the stock `IntimeModelSelector` widget accumulates the per-client validation metric each round and fires `AppEventType.GLOBAL_BEST_MODEL_AVAILABLE`, which `PTFileModelPersistor` already answers by saving `best_FL_global_model.pt`. That's the path NVFLARE ships for exactly this feature, and adopting it resolves the cluster of correctness issues below at once while removing ~150 lines of controller code (and the extra `get_persist_state`/`restore` surface). The current controller now sits as a thin subclass of stock `ScatterAndGather`, so the selector just listens to round events it already inherits — no controller changes needed at all.

Concrete concerns with the current approach, roughly in priority order:

**1. (blocking) The shipped default never actually selects a best model.** `app_files/config.json` sets `BEST_MODEL_METRIC: "VAL-F1-SCORE"`, but the client-API trainer only emits **per-lesion** labels `VAL-<METRIC>-<lesion>` (`trainer.py` ~L245, e.g. `VAL-F1-SCORE-Effusion`) and `VAL_LOSS` (~L238) — there is no bare `VAL-F1-SCORE`. `_track_validation_metric` gates on `label == self.best_model_metric` (`scatter_and_gather.py` ~L260), so the metric is never recorded, `_update_best_model` returns early every round, and `PersistToS3AndCleanup` takes its `shutil.copy(final → best)` fallback (`persist_and_cleanup.py` ~L149). Net result: the "best" model in every results zip is a byte-for-byte copy of the final model. I traced the metric flow end-to-end to confirm this (client `SummaryWriter` → `FlipAnalyticsBridge` → `SEND_RESULT` → controller), so it reproduces on the shipped tutorial as-is.

**2. The selection metric measures a different model than the one being saved.** Clients report `VAL_*` computed on their *locally-trained* models (after the local epochs); the controller then saves the *aggregated global* model for that round, judged by those client-local numbers. The score never describes the checkpoint actually written to disk. NVFLARE's convention (and `IntimeModelSelector`) is to score the *received global* model — clients evaluate it *before* local training and return it on `FLModel(metrics=...)` — so selection reflects the real global checkpoint.

**3. Cross-client selection is last-writer-wins, not an aggregate.** `self._round_metrics[self._current_round] = float(value)` (`scatter_and_gather.py` ~L274) overwrites on every client's metric event, so a round's value is whichever site reported last — one arbitrary site's validation split drives selection. `IntimeModelSelector` weight-averages across clients (optionally by local-iteration count).

**4. `best` and `final` files have different on-disk formats.** The final model is written by the persistor (`{model, train_conf}`); the best is a raw `torch.save(self._global_weights, ...)` (~L317), i.e. a `ModelLearnable` (`{weights, meta}`). A user downloading both needs different loading code per file. Routing the save through the persistor (which `GLOBAL_BEST_MODEL_AVAILABLE` does) yields identical formats.

**5. The no-selection fallback fabricates a "best".** Independently of #1, the `shutil.copy` fallback ships a copy of the final model labeled "best" when no selection ran — implying a selection that didn't happen. Cleaner to omit the file entirely when nothing was selected.

**6. The selection logic is untested.** `_track_validation_metric` / `_update_best_model` — the code deciding which checkpoint users receive — have no unit tests; the only test delta here is a moved-file count. Both are wrapped in broad `except Exception: log_error`, so a regression would fail silently.

**7. The production path isn't wired.** Only the tutorial `job.py` reads the keys; the fl-server's job assembly (`prepare_config`) injects config like `AGGREGATE_ONLY_REGEX` but has no equivalent for the best-model keys, so on the deployed platform the feature is currently inert (and, via #1, still ships a final-as-best copy).

I've put up #811 as a concrete implementation of the `IntimeModelSelector` approach — same `BEST_MODEL_METRIC` config UX, plus macro `VAL-<METRIC>` labels so the default actually resolves, verified end-to-end on the simulator (round 0 skipped, per-round cross-client averaging in the logs, best/final checkpoints with identical format and genuinely different weights). Happy to converge however is easiest — adopting #811, or refactoring this PR onto the stock components. The production-injection piece (#7) is a follow-up either way.

**Comment by atriaybagur** [comment]
I've put up https://github.com/londonaicentre/FLIP/pull/811 as a concrete implementation of the IntimeModelSelector approach — same BEST_MODEL_METRIC config UX, plus macro VAL-<METRIC> labels so the default actually resolves, verified end-to-end on the simulator (round 0 skipped, per-round cross-client averaging in the logs, best/final checkpoints with identical format and genuinely different weights). Happy to converge however is easiest — adopting https://github.com/londonaicentre/FLIP/pull/811, or refactoring this PR onto the stock components. The production-injection piece (https://github.com/londonaicentre/FLIP/issues/7) is a follow-up either way.
